From: Carlos Garnacho <carlosg@gnome.org>
Date: Thu, 5 Dec 2024 10:36:46 +0100
Subject: wayland: Handle reentrancy invalidating touch foci

The hashtable containing per-sequence touch info may change while
invalidating the focus, so grab a copy first of the existing sequences.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4166>

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/2025
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1966635
---
 src/wayland/meta-wayland-input.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/wayland/meta-wayland-input.c b/src/wayland/meta-wayland-input.c
index 5370c1a..a32a9d6 100644
--- a/src/wayland/meta-wayland-input.c
+++ b/src/wayland/meta-wayland-input.c
@@ -170,14 +170,15 @@ meta_wayland_input_invalidate_all_focus (MetaWaylandInput    *input,
 
   if (meta_wayland_seat_has_touch (seat))
     {
-      ClutterEventSequence *sequence;
+      g_autoptr (GList) touches = NULL;
+      GList *l;
 
       device = clutter_seat_get_pointer (clutter_seat);
       handler = wl_container_of (input->event_handler_list.next, handler, link);
 
-      g_hash_table_iter_init (&iter, seat->touch->touches);
-      while (g_hash_table_iter_next (&iter, (gpointer *) &sequence, NULL))
-        meta_wayland_event_handler_invalidate_focus (handler, device, sequence);
+      touches = g_hash_table_get_keys (seat->touch->touches);
+      for (l = touches; l; l = l->next)
+        meta_wayland_event_handler_invalidate_focus (handler, device, l->data);
     }
 
   if (meta_wayland_seat_has_touch (seat) &&
